[dynamo][user_defined] Replace UserFunctionVariable with VariableTracker build

anijain2305 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #165706

Audit: To prevent future issues with functools.partial or callable
objects.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

anijain2305 committed:
[dynamo][user_defined] Replace UserFunctionVariable with VariableTracker build

Audit: To prevent future issues with functools.partial or callable
objects.

[ghstack-poisoned]

Lucaskabela reviewed (approved):
nit: Maybe include some comment in the code on why we need to use build as opposed to the UserFunctionVariable here for posterity

Lucaskabela commented:
Otherwise LGTM!

clee2000 commented:
@pytorchbot revert -m "breaking internal tests D84961097" -c ghfirst

anijain2305 committed:
Update on "[dynamo][user_defined] Replace UserFunctionVariable with VariableTracker build"


Audit: To prevent future issues with functools.partial or callable
objects.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]
